vispy/app/backends/_tk.py

```diff
@@ -0,0 +1,790 @@
+    """
+    global _tk_inst, _tk_toplevels
+    tl = tk.Toplevel(_tk_inst, *args, **kwargs)
+    tl.withdraw()
+    _tk_toplevels.append(self)
+    return tl
+
+
+def _del_toplevel(fr=None):
+    """
+    Destroy the given Toplevel, and if it was the last one,
+    also destroy the global Tk instance if we created it.
+
+    :param fr: The CanvasBackend to destroy, defaults to None
+    """
+    global _tk_inst, _tk_inst_owned, _tk_toplevels
+
+    if fr:
+        try:
+            fr.destroy()
+            if fr.top:
+                fr.top.destroy()
+            _tk_toplevels.remove(fr)
+        except Exception:
+            pass
+
+    # If there are no Toplevels left, quit the mainloop.
+    if _tk_inst and not _tk_toplevels and _tk_inst_owned:
+        _tk_inst.quit()
+        _tk_inst.destroy()
+        _tk_inst = None
+
+
+class ApplicationBackend(BaseApplicationBackend):
+    def __init__(self):
+        BaseApplicationBackend.__init__(self)
+
+    def _vispy_get_backend_name(self):
+        return tk.__name__
+
+    def _vispy_process_events(self):
+        """Process events related to the spawned Tk application window.
+        First, update the global tk.Tk() instance, then call `_delayed_update`
+        on every created Toplevel (to force a redraw), and process some Tkinter
+        GUI events by calling the Tk.mainloop and immediately exiting.
+        """
+        # Update idle tasks first (probably not required)
+        app = self._vispy_get_native_app()
+        app.update_idletasks()
+
+        # Update every active Canvas window
+        for c in _tk_toplevels:
+            c._delayed_update()
+
+        # Process some events in the main Tkinter event loop
+        # And quit so we can continue elsewhere (call blocks normally)
+        app.after(0, lambda: app.quit())
+        app.mainloop()
+
+    def _vispy_run(self):
+        """Start the Tk.mainloop. This will block until all Tk windows are destroyed."""
+        self._vispy_get_native_app().mainloop()
+
+    def _vispy_quit(self):
+        """Destroy each created Toplevel by calling _vispy_close on it.
+        If there are no Toplevels left, also destroy the global Tk instance.
+        """
+        global _tk_toplevels
+        for c in _tk_toplevels:
+            c._vispy_close()
+        _del_toplevel()
+
+    def _vispy_get_native_app(self):
+        """Get or create the global Tk instance.
+
+        :return: Returns the instance.
+        """
+        global _tk_inst, _tk_inst_owned
+        if _tk_inst is None:
+            if tk._default_root:
+                # There already is a tk.Tk() instance available
+                _tk_inst = tk._default_root
+                _tk_inst_owned = False
+            else:
+                # Create our own top level Tk instance
+                _fix_tcl_lib()
+                _tk_inst = tk.Tk()
+                _tk_inst.withdraw()
+                _tk_inst_owned = True
+        return _tk_inst
```

Comment:
No there is only a global Tk() instance that every window uses as its parent if there is no other parent available.
This Tk() instance gets created the first time you create any tk related object, like a Toplevel but also just a Label or Button, because it's used to interface with the TCL interpreter that actually does the GUI stuff.

The thing is that if there is already an existing window that the OpenGL frame gets embedded in, you obviously don't want to close that one when closing the OpenGL frame, as that would end the application and that might be unintended.

On the other hand, if VisPy spawns one or more Tk windows, the first one would create the Tk instance and take it as parent, since it doesn't exist yet, and subsequent ones don't have to do that since they take the first one as parent and can just be created. If the user would now close the first window, all created windows would close, since that would destroy the base Tk() instance...
So to circumvent that, I resuse the Tk instance that is user-owned if available, or alternatively create my own (and hide it from view). Then when the last created window gets closed, [the global instance gets destroyed too](https://github.com/vispy/vispy/pull/1918/files/fb9889ebf78885b2d73f9e66f37f665a0727d9be#diff-db4b0a585243bf7475dda995d708b4d6bf0a7dd6b7cd813c5d5cef9ad4bacb47R288).

So to answer your question, Tk does not keep track, the very first thing that gets created, spaws the Tk instance, and when that gets destroyed, everything gets destroyed.

I don't know if it's a common pattern, in normal use you'd create a first window and that would own the instance. And when closing it, you'd expect your entire app to close with it. But since VisPy could spawn multiple windows, closing the first one would close everything.

Reply:
If you want to get rid of `global` statements, I could just encapsulate them inside a static class.

Reply:
I'm fine with globals in this case then. Thanks for the explanation.

Reply:
I already put them together with the classmethods `get_tk_instance`, `get_canvasses`, `new_toplevel` and `del_toplevel` in their own class `_TkInstanceManager`.